app/Http/Controllers/Rpc/WithdrawalController.php

```diff
@@ -21,11 +21,93 @@
 namespace Adshares\Adserver\Http\Controllers\Rpc;
 
 use Adshares\Adserver\Http\Controllers\Controller;
+use Adshares\Adserver\Utilities\AdsUtils;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Validator;
 
 class WithdrawalController extends Controller
 {
-    public function info()
+    /**
+     * Checks, if user has enough funds.
+     * @param $amount int transfer total amount
+     * @return bool true if has enough, false otherwise
+     */
+    public function hasUserEnoughFunds(int $amount): bool
     {
-        return self::json(json_decode('{"amount":10,"fee":9,"total":10}'));
+        // TODO check user account balance
+        $balance = 4000000000000000000;
+        return $amount <= $balance;
+    }
+
+    public function calculateWithdrawal(Request $request)
+    {
+        Validator::make($request->all(), [
+            'amount' => ['required', 'integer', 'min:1'],
+            'to' => 'required',
+        ])->validate();
+        $amount = $request->input('amount');
+        $addressTo = $request->input('to');
+
+        $addressFrom = config('app.adshares_address');
+        $fee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
+
+        if ($fee < 0) {
+            // invalid input for calculating fee
+            return self::json([], 422);
+        }
+        $total = $amount + $fee;
+        $resp = [
+            'amount' => $amount,
+            'fee' => $fee,
+            'total' => $total,
+        ];
+        return self::json($resp);
+    }
+
+    public function withdraw(Request $request)
+    {
+        Validator::make($request->all(), [
+            'amount' => ['required', 'integer', 'min:1'],
+            'to' => 'required',
+        ])->validate();
+
+        $amount = $request->input('amount');
+        $addressTo = $request->input('to');
+
+        $addressFrom = config('app.adshares_address');
+        $fee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
+
+        if ($fee < 0) {
+            // invalid input for calculating fee
+            return self::json([], 422);
+        }
+        $total = $amount + $fee;
+        if (!$this->hasUserEnoughFunds($total)) {
+            return self::json(['error' => 'not enough funds'], 400);
+        }
+
+        // TODO add tx to queue: $amount is amount, $addressTo is address
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)